fission-server/src/models/email_verification.rs

```diff
@@ -0,0 +1,154 @@
+    rng.gen_range(10000..=99999)
+}
+
+/// [Request] Parameters
+#[derive(Deserialize, Validate, Clone, Debug, ToSchema)]
+pub struct Request {
+    /// The email address of the user signing up
+    #[validate(email)]
+    pub email: String,
+    /// The (pre-generated) did of the client application.
+    /// Currently only did:key is supported.
+    pub did: String,
+    #[serde(skip)]
+    #[serde(default = "generate_code")]
+    code: u32,
+    /// The hash of the code, so that it can only be used by the intended recipient.
+    /// We only store the hash, not the code itself.
+    #[serde(skip_deserializing)]
+    pub code_hash: Option<String>,
+}
+
+impl Request {
+    /// Computes a hash of the code (so that it can only be used by the intended
+    /// recipient) and stores it in the struct.
+    pub fn compute_code_hash(&mut self) -> Result<()> {
+        if self.validate().is_err() {
+            log::error!("ERROR: Failed to validate the request.");
+            return Err(ValidationError::new("Failed to validate the request.").into());
+        }
+
+        let mut hasher = Sha256::new();
+        hasher.update(self.email.as_bytes());
+        hasher.update(self.did.as_bytes());
+        hasher.update(self.code.to_string().as_bytes());
+        let result = hasher.finish();
+        self.code_hash = Some(hex::encode(result));
+        Ok(())
+    }
+
+    /// Sends the code to the user
+    pub async fn send_code(&self) -> Result<()> {
+        if self.code_hash.is_none() {
+            log::error!("ERROR: Code hash must be generated before sending code.");
+            return Err(
+                ValidationError::new("Code hash must be generated before sending code.").into(),
+            );
+        }
+
+        let delivery_address = EmailAddress::address(&self.email.clone());
+
+        if self.validate().is_err() {
+            log::error!("ERROR: Failed to validate the request.");
+            return Err(ValidationError::new("Failed to validate the request.").into());
+        }
+
+        let server_settings = Settings::load()?;
+        let settings = server_settings.mailgun();
+
+        let mut template_vars = HashMap::new();
+        template_vars.insert("code".to_string(), self.code.to_string());
+
+        let message = Message {
+            to: vec![delivery_address],
+            subject: settings.subject.clone(),
+            template: settings.template.clone(),
+            template_vars,
+            ..Default::default()
+        };
+
+        print!("{} {}", settings.api_key.clone(), settings.domain.clone());
+        let client = Mailgun {
+            api_key: settings.api_key.clone(),
+            domain: settings.domain.clone(),
+            message,
+        };
+
+        let sender = EmailAddress::name_address(&settings.from_name, &settings.from_address);
+
+        if let Err(e) = tokio::task::spawn_blocking(move || client.send(&sender)).await? {
```

Comment:
My grasp on this is only tenuous because it has to do with tokio internals, but per https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html we need a separate process because the mailgun library uses a blocking call, and we can't block the main axum tokio process we're in. @zeeshanlakhani helped with this!